src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -187,16 +187,27 @@ static void uct_tcp_sockcm_ep_handle_remote_disconnect(uct_tcp_sockcm_ep_t *cep,
                                                        ucs_status_t status)
 {
     char peer_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t cb_status;
 
     /* remote peer disconnected */
-    ucs_debug("ep %p (fd=%d): remote peer (%s) disconnected (%s)", cep, cep->fd,
+    ucs_debug("ep %p (fd=%d): remote peer (%s) disconnected/rejected (%s)",
+              cep, cep->fd,
               uct_tcp_sockcm_cm_ep_peer_addr_str(cep, peer_str, UCS_SOCKADDR_STRING_LEN),
               ucs_status_string(status));
-    uct_tcp_sockcm_ep_reset_comm_ctx(cep);
 
-    cep->state |= UCT_TCP_SOCKCM_EP_GOT_DISCONNECT;
+    if (ucs_test_all_flags(cep->state, UCT_TCP_SOCKCM_EP_ON_CLIENT |
+                                       UCT_TCP_SOCKCM_EP_DATA_SENT) &&
+        !(cep->state & (UCT_TCP_SOCKCM_EP_HDR_RECEIVED |
+                        UCT_TCP_SOCKCM_EP_DATA_RECEIVED))) {
```

Comment:
should test CLIENT|UCT_TCP_SOCKCM_EP_DATA_RECEIVED
UCT_TCP_SOCKCM_EP_DATA_SENT|HDR_RECEIVED should be assertion

Reply:
after getting all the data and setting UCT_TCP_SOCKCM_EP_DATA_RECEIVED, we drop the UCT_TCP_SOCKCM_EP_HDR_RECEIVED flag -  https://github.com/openucx/ucx/blob/master/src/uct/tcp/tcp_sockcm_ep.c#L531

you meant to check that CLIENT is set and DATA_RECEIVED isn't, right? 
if yes, UCT_TCP_SOCKCM_EP_DATA_SENT doesnt have to be set yet

Reply:
so we need to check test_all_flags(CLINET|DATA_RECEIVED)
and if this is true, means we had to send all data, so assert(DATA_SENT)

Reply:
>so we need to check test_all_flags(CLINET|DATA_RECEIVED)

but on the client side, it's a reject if DATA_RECEIVED is not set yet.

Reply:
sorry, i've meant to test that CLINET is on, and DATA_RECEIVED is off

Reply:
so that matched a state before all data is sent too. if not all data is sent but got here then it's a disconnect, isn't it?

Reply:
i see. so we need to check all is sent, but not all received.
so, need to remove only UCT_TCP_SOCKCM_EP_HDR_RECEIVED check (and maybe make it an assertion)

Reply:
if just part of the data was received, which means that the UCT_TCP_SOCKCM_EP_HDR_RECEIVED bit is set but the UCT_TCP_SOCKCM_EP_DATA_RECEIVED is not, then this means that the server called uct_ep_create and therefore will not call uct_listener_reject. so if only part of the data arrived to the client, then it's a disconnect, not a reject, right? (even if it's some network disconnect, not by an api call)

Reply:
@yosefe wdyt?